diff --git a/node_modules/webpack-hot-middleware/client-overlay.js b/node_modules/webpack-hot-middleware/client-overlay.js
index 73e40c6..be4ced2 100644
--- a/node_modules/webpack-hot-middleware/client-overlay.js
+++ b/node_modules/webpack-hot-middleware/client-overlay.js
@@ -40,7 +40,9 @@ var entities = new Entities();
 
 function showProblems(type, lines) {
   clientOverlay.innerHTML = '';
-  lines.forEach(function(msg) {
+  lines.forEach(function(line) {
+    var isNested = typeof line === 'object';
+    var msg = isNested ? line.moduleName + '\n\n' + line.message : line;
     msg = ansiHTML(entities.encode(msg));
     var div = document.createElement('div');
     div.style.marginBottom = '26px';
diff --git a/node_modules/webpack-hot-middleware/client.js b/node_modules/webpack-hot-middleware/client.js
index abc95ed..b04b429 100644
--- a/node_modules/webpack-hot-middleware/client.js
+++ b/node_modules/webpack-hot-middleware/client.js
@@ -176,8 +176,10 @@ function createReporter() {
   var previousProblems = null;
   function log(type, obj) {
     var newProblems = obj[type]
-      .map(function(msg) {
-        return strip(msg);
+      .map(function(problem) {
+        var isNested = typeof problem === 'object';
+        var message = isNested ? problem.message : problem;
+        return strip(message);
       })
       .join('\n');
     if (previousProblems == newProblems) {
diff --git a/node_modules/webpack-hot-middleware/middleware.js b/node_modules/webpack-hot-middleware/middleware.js
index a5d7384..261a21a 100644
--- a/node_modules/webpack-hot-middleware/middleware.js
+++ b/node_modules/webpack-hot-middleware/middleware.js
@@ -122,6 +122,7 @@ function publishStats(action, statsResult, eventStream, log) {
     modules: true,
     timings: true,
     hash: true,
+    errors: true
   });
   // For multi-compiler, stats will be an object with a 'children' array of stats
   var bundles = extractBundles(stats);
