diff --git a/node_modules/chrome-aws-lambda/.DS_Store b/node_modules/chrome-aws-lambda/.DS_Store
new file mode 100644
index 0000000..a358587
Binary files /dev/null and b/node_modules/chrome-aws-lambda/.DS_Store differ
diff --git a/node_modules/chrome-aws-lambda/build/index.js b/node_modules/chrome-aws-lambda/build/index.js
index 5579560..1e10124 100644
--- a/node_modules/chrome-aws-lambda/build/index.js
+++ b/node_modules/chrome-aws-lambda/build/index.js
@@ -154,7 +154,7 @@ class Chromium {
             }
             return Promise.resolve('/tmp/chromium');
         }
-        const input = path_1.join(__dirname, '..', 'bin');
+        const input = path_1.join(process.cwd(),'node_modules',  'chrome-aws-lambda',  'bin');
         const promises = [
             lambdafs_1.default.inflate(`${input}/chromium.br`),
             lambdafs_1.default.inflate(`${input}/swiftshader.tar.br`),
